src/main/java/hudson/matrix/MatrixProject.java

```diff
@@ -660,6 +672,34 @@ public Combination apply(@Nullable List<String> strings) {
         return active;
     }
 
+    /**
+     * Configuration for matrix build
+     */
+    public static class RunConfiguration {
+        public Set<MatrixConfiguration> config;
+        public AxisList axisList;
+    }
+
+    /**
+     * Rebuild project settings and return actual configuration and axis list
+     */
+    /*package*/ RunConfiguration getRunConfiguration(MatrixBuildExecution context) throws IOException {
+        RunConfiguration runConfig = new RunConfiguration();
+        try {
+            buildLock.writeLock().lock();
```

Comment:
Maybe it is possible if the job is changed in the moment a build starts but because the way the axes are build I see no better solution.

I think the point that the axes are only rebuild if the build context is given may also helps to reduce the collisions.
Actual we are using the first bad implementation in our productive Jenkins without any problems since a half year.